In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
from surprise import Reader, Dataset, SVD, evaluate

import warnings; warnings.simplefilter('ignore')

## 1) Simple Recommender

In [2]:
md = pd. read_csv('the-movies-dataset/movies_metadata.csv')
print(md.columns)
md.head(2)


Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  ...  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1        Jumanji  When siblings Judy and Peter discover an encha...  ...   

  release_date      revenue runtime  \
0   1995-10-30  373554033.0    81.0   
1   1995-12-15  262797249.0   104.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   

                                     tagline      title  video vote_average  \
0                                        NaN  Toy Story  False          7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False          6.9   

  vote_count  
0     5415.0  
1     2413.0  

[2 rows x 24 columns]

In [3]:
md['genres'][0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [4]:
md['genres'].apply(lambda x: print(x))

[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10751, 'name': 'Family'}]
[{'id': 10749, 'name': 'Romance'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name

[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 28, 'name': 'Action'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}]
[]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'na

[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 12, 'name': 'Adventure'}]
[{'id': 10751, 'name': 'Family'}, {'id': 16, 'name': 'Animation'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 10751, 'name': 'Family'}, {'id': 16, 'name': 'Animation'}]
[{'id': 16, 'name': 'Animation'}, {'id': 10751, 'name': 'Family'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}, {'id': 14, 'name': 'Fantasy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10402, 'name':

[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}]
[]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 878, 'name': 'Science Fiction'}, {'id': 18, 'name': 'Drama'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 80, 'name': 'Crime'}, {'id': 35, 'name': 'Comedy'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 10749, 'name': 'Romance'}, {'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 28, 'name': 'Action'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 9648, 'name': 'Mystery'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]


[]
[{'id': 28, 'name': 'Action'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}, {'id': 10402, 'name': 'Music'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 16, 'name': 'Animation'}, {'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 28, 'name': 'Action'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 10752, 'name': 'War'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10402, 'name': 'Music'}]
[{'id':

[{'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'name': 'Adventure'}, {'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}, {'id': 10402, 'name': 'Music'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}, {'id': 10751, 'name': 'Family'}]
[{'id': 878, 'name': 'Science Fiction'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}, {'id': 80, 'name': 'Crime'}, {'id': 9648, 'name':

[{'id': 18, 'name': 'Drama'}, {'id': 10402, 'name': 'Music'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 10752, 'name': 'War'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 10752, 'name': 'War'}, {'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}, {'id': 18,

[{'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 27, 'name': 'Horror'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'T

[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 36, 'name': 'History'}, {'id': 18, 'name': 'Drama'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 37, 'name': 'Western'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 28, 'name': 'Action'}]

[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 10752, 'name': 'War'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 9648, 'name': 'Mystery'}, {'id': 10749, 'name': 'Romance'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 36, 'name': 'History'}, {'id': 99, 'name': 'Documentary'}, {'id': 10402, 'name': 'Music'}]
[{'id': 

[{'id': 10402, 'name': 'Music'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}, {'id': 80, 'name': 'Crime'}]
[{'id': 10749, 'name': 'Romance'}, {'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 80, 'name': 'Crime'}, {'id': 37, 'name': 'Western'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10402, 'name': 'Music'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10402, 'name': 'Music'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}, {'id': 10402, 'name': 'Music'}, {'id': 10749, '

[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}, {'id': 10752, 'name': 'War'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 99, 'name': 'Documentary'}, {'id': 18, 'name': 'Drama'}]
[{'id': 80, 'name': 'Crime'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 28, 'name': 'Action'}, {'id': 27, 'name': 'Horror'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 28, 'name': 'Action'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 28, 'name': 'Action'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18,

[{'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}, {'id': 10752, 'name': 'War'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 37, 'name': 'Western'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10402, 'name': 'Music'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 10752, 'name': 'War'}, {'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 14, 'name': 'Fantasy'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 18, 'name': 'Drama'}

[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 10752, 'name': 'War'}, {'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 16, 'name': 'Animation'}, {'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 16, 'name': 'Animation'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 36, 'name': 'History'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10402, 'name': 'Music'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id'

[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 9648, 'name': 'Mystery'}, {'id': 10749, 'name': 'Romance'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 99, 'name': 'Documentary'}, {'id': 36, 'name': 'History'}]
[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 16, 'name': 'Animation'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 16, 'name': 'Animation'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 10402, 'name': 'Music'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 10402, 'name': 'Music'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}

[{'id': 99, 'name': 'Documentary'}]
[{'id': 27, 'name': 'Horror'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 18, 'name': 'Drama'}, {'id': 37, 'name': 'Western'}, {'id': 80, 'name': 'Crime'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10752, 'name': 'War'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 10751, 'name': 'Family'}, {'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}, {'id': 27, 'name': 'Hor

[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 878, 'name': 'Science Fiction'}, {'id': 53, 'name': 'Thriller'}, {'id': 18, 'name': 'Drama'}, {'id': 80, 'name': 'Crime'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10770, 'name': 'TV Movie'}, {'id': 80, 'name': 'Crime'}]
[{'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}]
[]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 99, 'name': 'Documentary'}]
[]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 28, 'name': 'Action'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 27, 'name': 'Horror'

[]
[{'id': 16, 'name': 'Animation'}]
[{'id': 10751, 'name': 'Family'}]
[{'id': 16, 'name': 'Animation'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 37, 'name': 'Western'}]
[{'id': 18, 'name': 'Drama'}, {'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 28, 'name': 'Action'}, {'id': 18, 'name': 'Drama'}]
[]
[{'id': 16, 'name': 'Animation'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 99, 'name': 'Documentary'}]
[]
[{'id': 35, 'name': 'Comedy'}]
[]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 27, 'name': 'Horror'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name':

[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[]
[{'id': 99, 'name': 'Documentary'}, {'id': 10402, 'name': 'Music'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 10752, 'name': 'War'}, {'id': 99, 'name': 'Documentary'}]
[{'id': 28, 'name': 'Action'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 12, 'name': 'Adventure'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 10749, 'name': 'Romance'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10402, 'name': 'Music'}]
[{'id': 18, 'name': 'Drama'}, {'id': 80, 'name': 'Crime'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 99, 'name': 'Documentary'

[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10402, 'name': 'Music'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 10752, 'name': 'War'}, {'id': 18, 'name': 'Drama'}]
[{'id': 10752, 'name': 'War'}, {'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}, {'id': 28, 'name': 'Action'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 36, 'name': 'History'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 28, 'name': 'Action'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id'

[{'id': 10749, 'name': 'Romance'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10752, 'name': 'War'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10402, 'name': 'Music'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 80, 'name': 'Crime'}, {'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 10749, 'name': 'Romance'}, {'id': 80, 'name': 'Crime'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 12, 'na

[{'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 28, 'name': 'Action'}]
[{'id': 10751, 'name': 'Family'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}]
[{'id': 18, 'name': 'Drama'}, {'id': 14, 'name': 'Fantasy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 36, 'name': 'History'}, {'id': 10752, 'name': 'War'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10752, 'name': 'War'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}, {'id': 10752, 'name': 'War'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 53, 'name': 'Thriller'}, {'id': 80, 'name': 'Crime'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 99, 'name': 'Documentary'}, {'id': 10402, 'name': 'Music'}]
[{'id': 27, 'name': 'Horror'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 53, 'name': 'Thriller'}, {

[{'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 99, 'name': 'Documentary'}, {'id': 16, 'name': 'Animation'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 28, 'name': 'Action'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10402, 'name': 'Music'}]
[{'id': 18, 'name': 'Drama'}]
[]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 28, 'n

[{'id': 16, 'name': 'Animation'}]
[{'id': 99, 'name': 'Documentary'}]
[]
[{'id': 18, 'name': 'Drama'}, {'id': 10752, 'name': 'War'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 10402, 'name': 'Music'}, {'id': 99, 'name': 'Documentary'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 99, 'name': 'Documentary'}, {'id': 10402, 'name': 'Music'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 28, 'name': 'Action'}]
[{'id': 10752, 'name': 'War'}, {'id': 28, 'name': 'Action'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 28, 'name': 'Action'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 27, 'name': 'Horror'}, {'id': 9

[{'id': 53, 'name': 'Thriller'}, {'id': 27, 'name': 'Horror'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 99, 'name': 'Documentary'}, {'id': 14, 'name': 'Fantasy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 99, 'name': 'Documentary'}, {'id': 18, 'name': 'Drama'}]
[{'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 10752, 'name': 'War'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 878, 'name': 'Science Fiction'}]
[]
[{'id': 18, 'name': 'Drama'}]
[{'id': 80, 'name': 'Crime'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 99, 'name': 'Documentary'}, {'id': 10769, 'name': 'Foreign'}]

[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 27, 'name': 'Horror'}]
[{'id': 53, 'name': 'Thriller'}]
[{'id': 12, 'name': 'Adventure'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 16, 'name': 'Animation'}, {'id': 28, 'name': 'Action'}]
[{'id': 18, 'name': 'Drama'}, {'id': 80, 'name': 'Crime'}]
[{'id': 16, 'name': 'Animation'}, {'id': 12, 'name': 'Adventure'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 37, 'name': 'Western'}]
[{'id': 35, 'name': 'Com

[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 9648, 'name': 'Mystery'}, {'id': 80, 'name': 'Crime'}]
[{'id': 18, 'name': 'Drama'}, {'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 99, 'name': 'Documentary'}]
[{'id': 14, 'name': 'Fantasy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 80, 'name

[{'id': 18, 'name': 'Drama'}]
[{'id': 99, 'name': 'Documentary'}]
[]
[{'id': 16, 'name': 'Animation'}, {'id': 10751, 'name': 'Family'}]
[{'id': 16, 'name': 'Animation'}, {'id': 27, 'name': 'Horror'}, {'id': 9648, 'name': 'Mystery'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 16, 'name': 'Animation'}, {'id': 10751, 'name': 'Family'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 10751, 'name': 'Family'}, {'id': 16, 'name': 'Animation'}, {'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 27, 'name': 'Horror'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 28, 'name': 'Action'}, {'id': 27, 'name': 'Horror'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 18, 'name': 'Drama'}, {'id': 9648, 'name': 'Mystery'}, {'id': 80, 'name': 'Crime'}]
[{'id': 80, 'name': 

[{'id': 18, 'name': 'Drama'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 27, 'name': 'Horror'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 10770, 'name': 'TV Movie'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 9648, 'name': 'Mystery'}, {'id': 35, 'name': 'Comedy'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 9648, 'name': 'Mystery'}, {'id': 27, 'name': 'Horror'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 14, 'name': 'Fantasy'}, {'id': 28, 'name': 'Action'}, {'id': 18, 'name': 'Drama'}, {'id': 53, '

[{'id': 878, 'name': 'Science Fiction'}]
[{'id': 878, 'name': 'Science Fiction'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 10402, 'name': 'Music'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 10752, 'name': 'War'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 28, 'name': 'Action'}, {'id': 37, 'name': 'Western'}]
[{'id': 80, 'name': 'Crime'}]
[{'id': 37, 'name': 'Western'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 10770, 'name': 'TV Movie'}, {'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'

[{'id': 28, 'name': 'Action'}, {'id': 878, 'name': 'Science Fiction'}]
[]
[{'id': 10752, 'name': 'War'}, {'id': 99, 'name': 'Documentary'}]
[{'id': 28, 'name': 'Action'}, {'id': 18, 'name': 'Drama'}, {'id': 10752, 'name': 'War'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 9648, 'name': 'Mystery'}, {'id': 18, 'name': 'Drama'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 37, 'name': 'Western'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'n

[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 16, 'name': 'Animation'}, {'id': 12, 'name': 'Adventure'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 27, 'name': 'Horror'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 37, 'name': 'Western'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}]
[]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 878, 'name': 'Science Fiction'}, {'id': 28, 'name': 'Action'}, {'id': 14, 'name': 'Fantasy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 99, 'name': 'Documentary'}]
[{'id': 28, 'name': 'Action'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 80, 'name': 'Crime'}, {'

[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 14, 'name': 'Fantasy'}]
[]
[{'id': 12, 'name': 'Adventure'}, {'id': 53, 'name': 'Thriller'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 14, 'name': 'Fantasy'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 10751, 'name': 'Family'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 10751, 'name': 'Family'}, {'id': 16, 'name': 'Animation'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 14, 'name': 'Fantasy'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 16, 'name': 'Animation'}, {'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name':

[{'id': 18, 'name': 'Drama'}, {'id': 10770, 'name': 'TV Movie'}, {'id': 80, 'name': 'Crime'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 10770, 'name': 'TV Movie'}, {'id': 18, 'name': 'Drama'}, {'id': 27, 'name': 'Horror'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 10770, 'name': 'TV Movie'}, {'id': 18, 'name': 'Drama'}]
[{'id': 10770, 'name': 'TV Movie'}, {'id': 10751, 'name': 'Family'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}, {'id': 10770, 'name': 'TV Movie'}]
[{'id': 10751, 'name': 'Family'}]
[{'id': 10751, 'name': 'Family'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 14, 'name': 'Fantasy'}]
[{'id': 10770, 'name': 'TV Movie'}, {'id': 10751, 'name': 'Family'}, {'id': 35, 'name': 'Comedy'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'F

[{'id': 27, 'name': 'Horror'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 27, 'name': 'Horror'}, {'id': 14, 'name': 'Fantasy'}]
[]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 37, 'name': 'Western'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 35, 'name': 'Comedy'}]
[]
[{'id': 53, 'name': 'Thriller'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 10770, 'name': 'TV Movie'}, {'id': 27, 'name': 'Horror'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 16, 'name': 'Animation'}, {'id': 10751, 'name': 'Family'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 10749, 'name': 'Romance'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 27, 'name': 'Horror'}, {'id': 35, 'name': 'Comedy'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 53, 'name': 'Thriller'}]
[{'

[{'id': 12, 'name': 'Adventure'}, {'id': 10751, 'name': 'Family'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10402, 'name': 'Music'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 12, 'name': 'Adventure'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 99, 'name': 'Documentary'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 9648, 'name': 'Mystery'}, {'id': 12, 'name'

[{'id': 18, 'name': 'Drama'}, {'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 14, 'name': 'Fantasy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}]
[]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 27, 'name': 'Horror'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 53, 'name': 'Thriller'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}]
[{'id': 99,

[{'id': 99, 'name': 'Documentary'}]
[]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 18, 'name': 'Drama'}, {'id': 14, 'name': 'Fantasy'}]
[]
[{'id': 99, 'name': 'Documentary'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 28, 'name': 'Action'}]
[{'id': 18, 'name': 'Drama'}, {'id': 80, 'name': 'Crime'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[]
[{'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}]
[{'id': 10749

[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10769, 'name': 'Foreign'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[]
[{'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 18, 'name': 'Drama'}]
[{'id': 9648, 'name': 'Mystery'}, {'id': 27, 'name': 'Horror'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10402, 'name': 'Music'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 35, 'name': 'Comedy'}]
[]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 10751, 'name': 'Family'}, {'id':

[{'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}]
[{'id': 27, 'name': 'Horror'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 28, 'name': 'Action'}]
[{'id': 28, 'name': 'Action'}, {'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}, {'id': 53, 'name': 'Thriller'}, {'id': 10749, 'name': 'Romance'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 18, 'name': 'Drama'}, {'id': 16, 'name': 'Animation'}, {'id': 14, 'name': 'Fantasy'}]
[{'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 28, 'name': 'Action'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 35, 'name': 'Co

[{'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Action'}]
[]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[]
[{'id': 18, 'name': 'Drama'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 27, 'name': 'Horror'}, {'id': 18, 'name': 'Drama'}]
[]
[{'id': 53, 'name': 'Thriller'}, {'id': 80, 'name': 'Crime'}]
[{'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}]
[]
[{'id': 18, 'name': 'Drama'}, {'id': 12, 'name': 'Adventure'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 10752, 'name': 'War'}, {'id': 18, 'name': 'Drama'}]
[{'id': 99, 'name': 'Documentary'}]
[]
[{'id': 28, 'name': 'Action'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 10752, 'name': 'War'}, {'id': 36, 'name': 'History'}, {'id': 18, 'name': 'Drama'}]
[{'id': 16, 'name': 'Animation'}]
[{'id': 10751, 'name': 'Family'}, {'id': 16, 'name': 'Animation'}, {'

[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 16, 'name': 'Animation'}, {'id': 36, 'name': 'History'}, {'id': 18, 'name': 'Drama'}, {'id': 10752, 'name': 'War'}]
[{'id': 99, 'name': 'Documentary'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 878, 'name': 'Science Fiction'}]
[]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 10769, 'name': 'Foreign'}, {'id': 35, 'name': 'Comedy'}, {'id': 28, 'name': 'Action'}]
[{'id': 16, 'name': 'Animation'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 878, 'name': 'Science Fiction'}, {'id': 9648, 'name': 'Mystery'}, {'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 10751, 'name': 'Family'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 28, 'name

[{'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}]
[{'id': 10751, 'name': 'Family'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 10751, 'name': 'Family'}, {'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 10751, 'name': 'Family'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 27, 'name': 'Horror'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 27, 'name': 'Horror'}, {'id': 18, 'name': 'Drama'}]
[{'id': 10751, 'name': 'Family'}, {'id': 35, 'name': 'Comedy'}, {'id': 80, 'name': 'Crime'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'n

[{'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 28, 'name': 'Action'}, {'id': 80, 'name': 'Crime'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 80, 'name': 'Crime'}]
[{'id': 18, 'name': 'Drama'}, {'id': 36, 'name': 'History'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 14, 'name': 'Fantasy'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 27, 'name': 'Horror'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 16, 'name': 'Animation'}, {'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 16, 'name': 'Animation'

[{'id': 18, 'name': 'Drama'}, {'id': 10402, 'name': 'Music'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}, {'id': 37, 'name': 'Western'}]
[]
[{'id': 35, 'name': 'Comedy'}, {'id': 14, 'name': 'Fantasy'}, {'id': 27, 'name': 'Horror'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 9648, 'name': 'Mystery'}]
[{'id': 53, 'name': 'Thriller'}]
[{'id': 28, 'name': 'Action'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}]

[]
[{'id': 18, 'name': 'Drama'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 10752, 'name': 'War'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10751, 'name': 'Family'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 36, 'name': 'History'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 18, 'name': 'Drama'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10769, 'name': 'Foreign'}, {'id': 10749, 'name': 'Romance'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 878, 'name': 'Science Fiction'}]
[{'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 28, 'name': 'Action'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Roman

[{'id': 18, 'name': 'Drama'}]
[]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 53, 'name': 'Thriller'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 18, 'name': 'Drama'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}, {'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 28, 'name': 'Action'}, {'id': 16, 'name': 'Animation'}, {'id': 14, 'name': 'Fantasy'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 16, 'name': 'Animation'}, {'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 27, 'name': 'Horror'}, {'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'C

[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]
[{'id': 27, 'name': 'Horror'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'name': 'Adventure'}]
[{'id': 18, 'name': 'Drama'}, {'id': 878, 'name': 'Science Fiction'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 10769, 'name': 'Foreign'}]
[{'id': 80, 'name': 'Crime'}, {'id': 27, 'name': 'Horror'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 18, 'name': 'Drama'}, {'id': 878, 'name': 'Science Fiction'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}]
[{'id': 27, 'name': 'Horror'}]
[{'id': 27, 'name': 'Horror'}, {'id': 99, 'name': 'Documentary'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Come

[{'id': 53, 'name': 'Thriller'}]
[{'id': 99, 'name': 'Documentary'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 37, 'name': 'Western'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 80, 'name': 'Crime'}, {'id': 18, 'name': 'Drama'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 10752, 'name': 'War'}, {'id': 99, 'name': 'Documentary'}]
[{'id': 18, 'name': 'Drama'}]
[{'id': 99, 'name': 'Documentary'}]
[{'id': 80, 'name': 'Crime'}, {'id': 9648, 'name': 'Mystery'}]
[{'id': 18, 'name': 'Drama'}, {'id': 53, 'name': 'Thriller'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 35, 'name': 'Comedy'}]
[{'id': 18, 'name': 'Drama'}]
[]
[{'id': 878, 'name': 'Science Fiction'}, {'id': 53, 'name': 'Thriller'}, {'id': 18, 'name': 'Drama'}, {'id': 28, 'name': 'Ac

0        None
1        None
2        None
3        None
4        None
5        None
6        None
7        None
8        None
9        None
10       None
11       None
12       None
13       None
14       None
15       None
16       None
17       None
18       None
19       None
20       None
21       None
22       None
23       None
24       None
25       None
26       None
27       None
28       None
29       None
         ... 
45436    None
45437    None
45438    None
45439    None
45440    None
45441    None
45442    None
45443    None
45444    None
45445    None
45446    None
45447    None
45448    None
45449    None
45450    None
45451    None
45452    None
45453    None
45454    None
45455    None
45456    None
45457    None
45458    None
45459    None
45460    None
45461    None
45462    None
45463    None
45464    None
45465    None
Name: genres, Length: 45466, dtype: object

In [5]:
md['genres'] = md['genres'].fillna('[]').apply(literal_eval)
md['genres'][0]

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

In [6]:
md['genres'] = md['genres'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
md['genres'][:5]

0     [Animation, Comedy, Family]
1    [Adventure, Fantasy, Family]
2               [Romance, Comedy]
3        [Comedy, Drama, Romance]
4                        [Comedy]
Name: genres, dtype: object

**I use the TMDB Ratings to come up with our Top Movies Chart. I will use IMDB's weighted rating formula to construct my chart. Mathematically, it is represented as follows: Weighted Rating, (WR) = (vv+m.R)+(mv+m.C)**

where,

-  'v' is the number of votes for the movie
-  'm' is the minimum votes required to be listed in the chart
-  'R' is the average rating of the movie
-  'C' is the mean vote across the whole report

**The next step is to determine an appropriate value for 'm'**, the minimum votes required to be listed in the chart. We will use 95th percentile as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.

**I will build our overall Top 250 Chart and will define a function to build charts for a particular genre. Let's begin!**

In [7]:
vote_averages = md[md['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
C

5.244896612406511

In [8]:
#For a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list,
vote_counts = md[md['vote_count'].notnull()]['vote_count'].astype('int')
m = vote_counts.quantile(0.95)
m

434.0

Therefore, to qualify to be considered for the chart, a movie has to have at least 434 votes on TMDB.

In [9]:
md['year'] = pd.to_datetime(md['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [10]:
qualified = md[(md['vote_count'] >= m) & (md['vote_count'].notnull()) & (md['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified.shape

(2274, 6)

Therefore, to qualify to be considered for the chart, a movie has to have at least 434 votes on TMDB. We also see that the average rating for a movie on TMDB is 5.244 on a scale of 10. 
>2274 Movies qualify to be on our chart.

In [11]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [12]:
qualified['wr'] = qualified.apply(weighted_rating, axis=1)

In [13]:
qualified = qualified.sort_values('wr', ascending=False).head(250)

### Top Movies

In [14]:
qualified.head(5)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
22879                                       Interstellar  2014       11187   
2843                                          Fight Club  1999        9678   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   

       vote_average popularity  \
15480             8    29.1081   
12481             8    123.167   
22879             8    32.2135   
2843              8    63.8696   
4863              8    32.0707   

                                                  genres        wr  
15480  [Action, Thriller, Science Fiction, Mystery, A...  7.917588  
12481                   [Drama, Action, Crime, Thriller]  7.905871  
22879                [Adventure, Drama, Science Fiction]  7.897107  
2843                                             [Drama]  7.881753  
4863                        [Adventure, Fantasy, Action]  7.871787

In [69]:
s = md.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'genre'
print(s)
gen_md = md.drop('genres', axis=1).join(s)

0              Animation
0                 Comedy
0                 Family
1              Adventure
1                Fantasy
1                 Family
2                Romance
2                 Comedy
3                 Comedy
3                  Drama
3                Romance
4                 Comedy
5                 Action
5                  Crime
5                  Drama
5               Thriller
6                 Comedy
6                Romance
7                 Action
7              Adventure
7                  Drama
7                 Family
8                 Action
8              Adventure
8               Thriller
9              Adventure
9                 Action
9               Thriller
10                Comedy
10                 Drama
              ...       
46608             Comedy
46611          Animation
46611             Family
46612            Fantasy
46612             Action
46612           Thriller
46613             Comedy
46613            Fantasy
46614        Documentary


In [16]:
def build_chart(genre, percentile=0.85):
    df = gen_md[gen_md['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    
    return qualified

### Top Romance Movies

In [17]:
build_chart('Romance').head(5)

title  year  vote_count  vote_average popularity  \
10309  Dilwale Dulhania Le Jayenge  1995         661             9     34.457   
351                   Forrest Gump  1994        8147             8    48.3072   
876                        Vertigo  1958        1162             8    18.2082   
40251                   Your Name.  2016        1030             8  34.461252   
883               Some Like It Hot  1959         835             8    11.8451   

             wr  
10309  8.565285  
351    7.971357  
876    7.811667  
40251  7.789489  
883    7.745154

## 2) Content Based Recommender


In [18]:
links_small = pd.read_csv('the-movies-dataset/links_small.csv')
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')
print(links_small.head())

0      862
1     8844
2    15602
3    31357
4    11862
Name: tmdbId, dtype: int64


In [19]:
md = md.drop([19730, 29503, 35587])

In [20]:
#Check EDA Notebook for how and why I got these indices.
md['id'] = md['id'].astype('int')

In [21]:
smd = md[md['id'].isin(links_small)]
smd.shape

(9099, 25)

We have 9099 movies avaiable in our small movies metadata dataset which is 5 times smaller than our original dataset of 45000 movies.

### Movie Description Based Recommender

Let us first try to build a recommender using movie descriptions and taglines. 

In [22]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

In [23]:
tf = TfidfVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
tfidf_matrix = tf.fit_transform(smd['description'])

In [24]:
tfidf_matrix.shape

(9099, 268124)

In [25]:
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)
cosine_sim[0]

array([1.        , 0.00680476, 0.        , ..., 0.        , 0.00344913,
       0.        ])

We now have a pairwise cosine similarity matrix for all the movies in our dataset. The next step is to write a function that returns the 30 most similar movies based on the cosine similarity score.

In [27]:
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return titles.iloc[movie_indices]

We're all set. Let us now try and get the top recommendations for a few movies and see how good the recommendations are.

In [28]:
get_recommendations('The Godfather').head(10)

973      The Godfather: Part II
8387                 The Family
3509                       Made
4196         Johnny Dangerously
29               Shanghai Triad
5667                       Fury
2412             American Movie
1582    The Godfather: Part III
4221                    8 Women
2159              Summer of Sam
Name: title, dtype: object

In [29]:
get_recommendations('The Dark Knight').head(10)

7931                      The Dark Knight Rises
132                              Batman Forever
1113                             Batman Returns
8227    Batman: The Dark Knight Returns, Part 2
7565                 Batman: Under the Red Hood
524                                      Batman
7901                           Batman: Year One
2579               Batman: Mask of the Phantasm
2696                                        JFK
8165    Batman: The Dark Knight Returns, Part 1
Name: title, dtype: object

In [30]:
credits = pd.read_csv('the-movies-dataset/credits.csv')
keywords = pd.read_csv('the-movies-dataset/keywords.csv')

In [31]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
md['id'] = md['id'].astype('int')
md.shape

(45463, 25)

In [32]:
md = md.merge(credits, on='id')
md = md.merge(keywords, on='id')

In [33]:
smd = md[md['id'].isin(links_small)]
smd.shape

(9219, 28)

In [34]:
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x))
smd['crew_size'] = smd['crew'].apply(lambda x: len(x))

In [35]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [36]:
smd['director'] = smd['crew'].apply(get_director)

In [37]:
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
smd['cast'] = smd['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)

In [38]:
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

In [39]:
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [40]:
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
smd['director'] = smd['director'].apply(lambda x: [x,x, x])

Keywords
We will do a small amount of pre-processing of our keywords before putting them to any use. As a first step, we calculate the frequenct counts of every keyword that appears in the dataset.

In [41]:
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [42]:
s = s.value_counts()
s[:5]

independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: keyword, dtype: int64

Keywords occur in frequencies ranging from 1 to 610. We do not have any use for keywords that occur only once. Therefore, these can be safely removed. Finally, we will convert every word to its stem so that words such as Dogs and Dog are considered the same.

In [43]:
s = s[s > 1]

In [44]:
stemmer = SnowballStemmer('english')
stemmer.stem('dogs')

'dog'

In [45]:
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words

In [46]:
smd['keywords'] = smd['keywords'].apply(filter_keywords)
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

smd['soup'] = smd['keywords'] + smd['cast'] + smd['director'] + smd['genres']
smd['soup'] = smd['soup'].apply(lambda x: ' '.join(x))


In [47]:
count = CountVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [48]:
cosine_sim = cosine_similarity(count_matrix, count_matrix)

In [49]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [50]:
get_recommendations('The Dark Knight').head(10)

8031         The Dark Knight Rises
6218                 Batman Begins
6623                  The Prestige
2085                     Following
7648                     Inception
4145                      Insomnia
3381                       Memento
8613                  Interstellar
7659    Batman: Under the Red Hood
1134                Batman Returns
Name: title, dtype: object

In [51]:
get_recommendations('Mean Girls').head(10)

3319               Head Over Heels
4763                 Freaky Friday
1329              The House of Yes
6277              Just Like Heaven
7905         Mr. Popper's Penguins
7332    Ghosts of Girlfriends Past
6959     The Spiderwick Chronicles
8883                      The DUFF
6698         It's a Boy Girl Thing
7377       I Love You, Beth Cooper
Name: title, dtype: object

In [52]:
def improved_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year']]
    vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(0.60)
    qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    qualified['wr'] = qualified.apply(weighted_rating, axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(10)
    return qualified

In [53]:
improved_recommendations('The Dark Knight')

title  vote_count  vote_average  year  \
7648                           Inception       14075             8  2010   
8613                        Interstellar       11187             8  2014   
6623                        The Prestige        4510             8  2006   
3381                             Memento        4168             8  2000   
8031               The Dark Knight Rises        9263             7  2012   
6218                       Batman Begins        7511             7  2005   
1134                      Batman Returns        1706             6  1992   
132                       Batman Forever        1529             5  1995   
9024  Batman v Superman: Dawn of Justice        7189             5  2016   
1260                      Batman & Robin        1447             4  1997   

            wr  
7648  7.917588  
8613  7.897107  
6623  7.758148  
3381  7.740175  
8031  6.921448  
6218  6.904127  
1134  5.846862  
132   5.054144  
9024  5.013943  
1260  4.287233

Let me also get the recommendations for Mean Girls, my girlfriend's favorite movie.

In [54]:
improved_recommendations('Mean Girls')

title  vote_count  vote_average  year  \
1547                       The Breakfast Club        2189             7  1985   
390                        Dazed and Confused         588             7  1993   
8883                                 The DUFF        1372             6  2015   
3712                     The Princess Diaries        1063             6  2001   
4763                            Freaky Friday         919             6  2003   
6277                         Just Like Heaven         595             6  2005   
6959                The Spiderwick Chronicles         593             6  2008   
7494  American Pie Presents: The Book of Love         454             5  2009   
7332               Ghosts of Girlfriends Past         716             5  2009   
7905                    Mr. Popper's Penguins         775             5  2011   

            wr  
1547  6.709602  
390   6.254682  
8883  5.818541  
3712  5.781086  
4763  5.757786  
6277  5.681521  
6959  5.680901  
7494  5.119690  
7332  5.092422  
7905  5.087912

## 3) Collaborative Filtering

In [55]:
reader = Reader()

In [56]:
ratings = pd.read_csv('the-movies-dataset/ratings_small.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [57]:
data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)
data.split(n_folds=5)

In [58]:
svd = SVD()
evaluate(svd, data, measures=['RMSE', 'MAE'])

Evaluating RMSE, MAE of algorithm SVD.

------------
Fold 1
RMSE: 0.9020
MAE:  0.6947
------------
Fold 2
RMSE: 0.9054
MAE:  0.6959
------------
Fold 3
RMSE: 0.8992
MAE:  0.6929
------------
Fold 4
RMSE: 0.8952
MAE:  0.6901
------------
Fold 5
RMSE: 0.8855
MAE:  0.6826
------------
------------
Mean RMSE: 0.8975
Mean MAE : 0.6912
------------
------------


CaseInsensitiveDefaultDict(list,
                           {'rmse': [0.901979924741404,
                             0.9054256182348522,
                             0.8992228093212598,
                             0.8951782612721974,
                             0.8855392709812152],
                            'mae': [0.694725153372281,
                             0.6958539318217645,
                             0.6928835419976481,
                             0.6900867377691142,
                             0.6826210782623813]})

We get a mean Root Mean Sqaure Error of 0.8963 which is more than good enough for our case. Let us now train on our dataset and arrive at predictions.

In [59]:
trainset = data.build_full_trainset()
svd.train(trainset)

Let us pick user 5000 and check the ratings s/he has given.

In [60]:
ratings[ratings['userId'] == 1]

userId  movieId  rating   timestamp
0        1       31     2.5  1260759144
1        1     1029     3.0  1260759179
2        1     1061     3.0  1260759182
3        1     1129     2.0  1260759185
4        1     1172     4.0  1260759205
5        1     1263     2.0  1260759151
6        1     1287     2.0  1260759187
7        1     1293     2.0  1260759148
8        1     1339     3.5  1260759125
9        1     1343     2.0  1260759131
10       1     1371     2.5  1260759135
11       1     1405     1.0  1260759203
12       1     1953     4.0  1260759191
13       1     2105     4.0  1260759139
14       1     2150     3.0  1260759194
15       1     2193     2.0  1260759198
16       1     2294     2.0  1260759108
17       1     2455     2.5  1260759113
18       1     2968     1.0  1260759200
19       1     3671     3.0  1260759117

In [61]:
svd.predict(1, 302, 3)

Prediction(uid=1, iid=302, r_ui=3, est=2.62543233704181, details={'was_impossible': False})

## 4) Hybrid Recommender

In [62]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [63]:
id_map = pd.read_csv('the-movies-dataset/links_small.csv')[['movieId', 'tmdbId']]
id_map['tmdbId'] = id_map['tmdbId'].apply(convert_int)
id_map.columns = ['movieId', 'id']
id_map = id_map.merge(smd[['title', 'id']], on='id').set_index('title')
#id_map = id_map.set_index('tmdbId')

In [64]:
indices_map = id_map.set_index('id')

In [65]:
def hybrid(userId, title):
    idx = indices[title]
    tmdbId = id_map.loc[title]['id']
    #print(idx)
    movie_id = id_map.loc[title]['movieId']
    
    sim_scores = list(enumerate(cosine_sim[int(idx)]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year', 'id']]
    movies['est'] = movies['id'].apply(lambda x: svd.predict(userId, indices_map.loc[x]['movieId']).est)
    movies = movies.sort_values('est', ascending=False)
    return movies.head(10)

In [66]:
hybrid(1, 'Avatar')

title  vote_count  vote_average  year  \
974                               Aliens      3282.0           7.7  1986   
522           Terminator 2: Judgment Day      4274.0           7.7  1991   
1011                      The Terminator      4208.0           7.4  1984   
8658          X-Men: Days of Future Past      6155.0           7.5  2014   
2014                    Fantastic Planet       140.0           7.6  1973   
8401             Star Trek Into Darkness      4479.0           7.4  2013   
1376                             Titanic      7770.0           7.5  1997   
922                            The Abyss       822.0           7.1  1989   
344                            True Lies      1138.0           6.8  1994   
1621  Darby O'Gill and the Little People        35.0           6.7  1959   

          id       est  
974      679  3.256587  
522      280  3.225207  
1011     218  3.081341  
8658  127585  2.883082  
2014   16306  2.854690  
8401   54138  2.798777  
1376     597  2.742528  
922     2756  2.720767  
344    36955  2.699811  
1621   18887  2.684168

In [67]:
hybrid(500, 'Avatar')

title  vote_count  vote_average  year  \
2014                    Fantastic Planet       140.0           7.6  1973   
8401             Star Trek Into Darkness      4479.0           7.4  2013   
1011                      The Terminator      4208.0           7.4  1984   
522           Terminator 2: Judgment Day      4274.0           7.7  1991   
8658          X-Men: Days of Future Past      6155.0           7.5  2014   
1621  Darby O'Gill and the Little People        35.0           6.7  1959   
2132                         Superman II       642.0           6.5  1980   
1376                             Titanic      7770.0           7.5  1997   
4017                     Hawk the Slayer        13.0           4.5  1980   
7265                Dragonball Evolution       475.0           2.9  2009   

          id       est  
2014   16306  3.369546  
8401   54138  3.262859  
1011     218  3.261427  
522      280  3.229205  
8658  127585  3.209830  
1621   18887  3.163991  
2132    8536  3.056952  
1376     597  3.043459  
4017   25628  3.015466  
7265   14164  2.986367